### PR TITLE
Move the Critic into the Optimizer

### PR DESCRIPTION
**Pull request by andrewcoh**
### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/action_info.py [hunk 1]
```diff
@@ -8,10 +8,9 @@
 class ActionInfo(NamedTuple):
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/action_info.py:8]: I know this is not part of this PR, but can you add a comment for each of the entries on the ActionInfo tuple ?

File: ml-agents/mlagents/trainers/ppo/optimizer_torch.py [hunk 2]
```diff
@@ -25,7 +26,21 @@ def __init__(self, policy: TorchPolicy, trainer_settings: TrainerSettings):
         # Create the graph here to give more granular control of the TF graph to the Optimizer.
 
         super().__init__(policy, trainer_settings)
-        params = list(self.policy.actor_critic.parameters())
+        reward_signal_configs = trainer_settings.reward_signals
+        reward_signal_names = [key.value for key, _ in reward_signal_configs.items()]
+
+        if policy.shared_critic:
+            self.value_net = policy.actor
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/ppo/optimizer_torch.py:33]: rename to _value_net (private property) or even self._critic since it is used as critic.

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 3]
```diff
@@ -483,7 +495,7 @@ def update(self, batch: AgentBuffer, num_sequences: int) -> Dict[str, float]:
         offset = 1 if self.policy.sequence_length > 1 else 0
         next_memories_list = [
             ModelUtils.list_to_tensor(
-                batch[BufferKey.MEMORY][i][self.policy.m_size // 2 :]
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:486]: Are there changes to memories in this PR? Why was this line changed?

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py [hunk 4]
```diff
@@ -565,7 +581,11 @@ def update(self, batch: AgentBuffer, num_sequences: int) -> Dict[str, float]:
         policy_loss = self.sac_policy_loss(log_probs, q1p_out, masks)
         entropy_loss = self.sac_entropy_loss(log_probs, masks)
 
-        total_value_loss = q1_loss + q2_loss + value_loss
+        if self.policy.shared_critic:
+            policy_loss += value_loss
+            total_value_loss = q1_loss + q2_loss
+        else:
+            total_value_loss = q1_loss + q2_loss + value_loss
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/sac/optimizer_torch.py:588]:
```suggestion
        total_value_loss = q1_loss + q2_loss
        if self.policy.shared_critic:
            policy_loss += value_loss
        else:
            total_value_loss += value_loss
```

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 5]
```diff
@@ -156,7 +156,31 @@ def forward(
...
+        """
+        Get value outputs for the given obs.
+        :param inputs: List of inputs as tensors.
+        :param memories: Tensor of memories, if using memory. Otherwise, None.
+        :returns: Dict of reward stream to output tensor for values.
+        """
+        pass
+
+
+class ValueNetwork(nn.Module, Critic):
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/networks.py:183]: Maybe later, but we could rename SimpleCritic to mirror SimpleActor

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 6]
```diff
@@ -156,7 +156,31 @@ def forward(
         return encoding, memories
 
 
-class ValueNetwork(nn.Module):
+class Critic(abc.ABC):
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/networks.py:159]: I think it is time we split networks.py into networks.py, actor.py and Critic.py. Maybe in a subsequent PR.

**Review by vincentpierre — APPROVED** [review]

File: ml-agents/mlagents/trainers/sac/optimizer_torch.py (hunk 3 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/sac/optimizer_torch.py:486]: https://github.com/Unity-Technologies/ml-agents/pull/4948

File: ml-agents/mlagents/trainers/policy/torch_policy.py [hunk 7]
```diff
@@ -61,31 +57,40 @@ def __init__(
...
+            self.shared_critic = True
+
         # Save the m_size needed for export
         self._export_m_size = self.m_size
         # m_size needed for training is determined by network, not trainer settings
-        self.m_size = self.actor_critic.memory_size
+        self.m_size = self.actor.memory_size
 
-        self.actor_critic.to(default_device())
+        self.actor.to(default_device())
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/policy/torch_policy.py:93]: Do we need to move the critic to the default device too?

File: ml-agents/mlagents/trainers/ppo/optimizer_torch.py [hunk 8]
```diff
@@ -25,7 +26,19 @@ def __init__(self, policy: TorchPolicy, trainer_settings: TrainerSettings):
...
+
+        if policy.shared_critic:
+            self._critic = policy.actor
+        else:
+            self._critic = ValueNetwork(
+                reward_signal_names,
+                policy.behavior_spec.observation_specs,
+                network_settings=trainer_settings.network_settings,
+            )
+
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/ppo/optimizer_torch.py:40]: I think we need to move the critic to `default_device` like we did in the Policy

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py [hunk 9]
```diff
@@ -49,25 +55,115 @@ def create_reward_signals(self, reward_signal_configs):
...
+
     def get_trajectory_value_estimates(
-        self, batch: AgentBuffer, next_obs: List[np.ndarray], done: bool
-    ) -> Tuple[Dict[str, np.ndarray], Dict[str, float]]:
+        self,
+        batch: AgentBuffer,
+        next_obs: List[np.ndarray],
+        done: bool,
+        agent_id: str = "",
+    ) -> Tuple[Dict[str, np.ndarray], Dict[str, float], Optional[AgentBufferField]]:
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:130]: docstring

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py [hunk 10]
```diff
@@ -49,25 +55,115 @@ def create_reward_signals(self, reward_signal_configs):
...
+        _mem = initial_memory
+        seq_obs = []
+
+        first_seq_len = self.policy.sequence_length
+        for _obs in tensor_obs:
+            if leftover > 0:
+                first_seq_obs = _obs[0:leftover]
+                first_seq_len = leftover
+            else:
+                first_seq_obs = _obs[0 : self.policy.sequence_length]
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:82]:
```suggestion
            if leftover > 0:
                first_seq_len = leftover
            first_seq_obs = _obs[0 :first_seq_len]
```

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py [hunk 11]
```diff
@@ -49,25 +55,115 @@ def create_reward_signals(self, reward_signal_configs):
                 reward_signal, self.policy.behavior_spec, settings
             )
 
+    def _evaluate_by_sequence(
+        self, tensor_obs: List[torch.Tensor], initial_memory: np.ndarray
+    ) -> Tuple[Dict[str, torch.Tensor], AgentBufferField, torch.Tensor]:
+        """
+        Evaluate the batch sequence-by-sequence, assembling the result. This enables us to get the
+        intermediate memories for the critic.
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:63]: describe the inputs and their dimensions. it is not clear what initial_memory's shape is

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py [hunk 12]
```diff
@@ -49,25 +55,115 @@ def create_reward_signals(self, reward_signal_configs):
...
+        for _obs in tensor_obs:
+            if leftover > 0:
+                first_seq_obs = _obs[0:leftover]
+                first_seq_len = leftover
+            else:
+                first_seq_obs = _obs[0 : self.policy.sequence_length]
+            seq_obs.append(first_seq_obs)
+
+        for _ in range(first_seq_len):
+            all_next_memories.append(initial_memory.squeeze().detach().numpy())
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:86]:
```suggestion
            all_next_memories.append(_mem.squeeze().detach().numpy())
```

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py [hunk 13]
```diff
@@ -49,25 +55,115 @@ def create_reward_signals(self, reward_signal_configs):
...
+        # In the buffer, the 1st sequence are the ones that are padded. So if seq_len = 3 and
+        # trajectory is of length 10, the 1st sequence is [pad,pad,obs].
+        # Compute the number of elements in this padded seq.
+        leftover = num_experiences % self.policy.sequence_length
+
+        # Compute values for the potentially truncated initial sequence
+        _mem = initial_memory
+        seq_obs = []
+
+        first_seq_len = self.policy.sequence_length
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:76]: Any reason for padding the end of the first sequence and not the last sequence ?

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py (hunk 12 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:86]: Hard to understand why the all_next_memories are a concatenation of initial memories...

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py (hunk 13 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:76]: Neither is padded in this function, but the buffer pads the front of the first sequence, so this function accounts for that. I left the padding in the buffer as-is and still need to experiment with that change; it will likely be a future PR.

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py (hunk 12 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:86]: Refactored this to be a bit clearer

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py (hunk 11 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:63]: Added a comment

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py (hunk 9 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:136]: Added

File: ml-agents/mlagents/trainers/torch/networks.py [hunk 14]
```diff
@@ -358,6 +370,21 @@ def get_action_stats(
         action, log_probs, entropies = self.action_model(encoding, masks)
         return action, log_probs, entropies, memories
 
+    def get_stats(
```
[the remark this answers is not in the document]
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/networks.py:373]: docstring in the interface seems to be what we have done in the past

[the remark this answers is not in the document]
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/torch/networks.py:373]: https://github.com/Unity-Technologies/ml-agents/pull/4939/files#diff-50a682c59106eecab8f1dd4c6ef6152f02bd52ee127074a225f1c8b52b68278eR273

File: ml-agents/mlagents/trainers/optimizer/torch_optimizer.py (hunk 13 above)
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/optimizer/torch_optimizer.py:81]: Then why is the first sequence handled differently ?

File: ml-agents/mlagents/trainers/torch/networks.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (vincentpierre) [reply ml-agents/mlagents/trainers/torch/networks.py:373]: ok, I forgot this was implementing the interface

**Review by ervteng — APPROVED** [review]
:shipit:
